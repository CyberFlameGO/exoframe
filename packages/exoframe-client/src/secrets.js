import got from 'got';

export const listSecrets = async ({ endpoint, token }) => {
  const remoteUrl = `${endpoint}/secrets`;

  // construct shared request params
  const options = {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token}`,
    },
    responseType: 'json',
  };

  // try sending request
  try {
    const {
      body: { secrets },
    } = await got(remoteUrl, options);
    return secrets;
  } catch (e) {
    // if authorization is expired/broken/etc
    if (e.response.statusCode === 401) {
      throw new Error('Authorization expired!');
    }

    throw e;
  }
};

export const createSecret = async ({ name, value, endpoint, token }) => {
  const remoteUrl = `${endpoint}/secrets`;

  // construct shared request params
  const options = {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
    },
    responseType: 'json',
    json: {
      secretName: name,
      secretValue: value,
    },
  };
  // try sending request
  try {
    const { body } = await got(remoteUrl, options);
    return body;
  } catch (e) {
    // if authorization is expired/broken/etc
    if (e.response.statusCode === 401) {
      throw new Error('Authorization expired!');
    }

    throw e;
  }
};

export const getSecret = async ({ name, endpoint, token }) => {
  const remoteUrl = `${endpoint}/secrets`;

  // construct shared request params
  const options = {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token}`,
    },
    responseType: 'json',
  };

  // try sending request
  try {
    const {
      body: { secret },
    } = await got(`${remoteUrl}/${name}`, options);
    return secret;
  } catch (e) {
    // if authorization is expired/broken/etc
    if (e.response.statusCode === 401) {
      throw new Error('Authorization expired!');
    }

    throw e;
  }
};

export const removeSecret = async ({ name, endpoint, token }) => {
  const remoteUrl = `${endpoint}/secrets`;

  // construct shared request params
  const rmOptions = {
    method: 'DELETE',
    headers: {
      Authorization: `Bearer ${token}`,
    },
    responseType: 'json',
    json: {
      secretName: name,
    },
  };
  try {
    const { body, statusCode } = await got(remoteUrl, rmOptions);
    if (statusCode !== 204) {
      throw new Error(`Error removing deployment secret: ${body.reason || 'Please try again!'}`);
    }

    return true;
  } catch (e) {
    // if authorization is expired/broken/etc
    if (e.response.statusCode === 401) {
      throw new Error('Authorization expired!');
    }

    throw e;
  }
};
