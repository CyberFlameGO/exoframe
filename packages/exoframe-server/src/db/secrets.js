// npm packages
import Loki from 'lokijs';
import { join } from 'path';
// our packages
import { baseFolder } from '../config/index.js';

// init in-memory adapter for login requests
const memAdapter = new Loki.LokiMemoryAdapter();
const fsAdapter = new Loki.LokiFsAdapter();

// init persistent secrets db
let secretsCollection = {};
let secretResolve = () => {};
export const secretsInited = new Promise((r) => {
  secretResolve = r;
});
export const secretDb = new Loki(join(baseFolder, 'secrets.db'), {
  adapter: process.env.NODE_ENV !== 'testing' ? fsAdapter : memAdapter,
  autoload: true,
  autoloadCallback() {
    // get of create secrets collection
    secretsCollection = secretDb.getCollection('secrets');
    if (secretsCollection === null) {
      secretsCollection = secretDb.addCollection('secrets');
    }
    secretResolve();
  },
  autosave: process.env.NODE_ENV !== 'testing',
});

export function getSecretsCollection() {
  return secretsCollection;
}
