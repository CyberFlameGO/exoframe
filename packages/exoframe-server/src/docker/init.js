// our modules
import { getConfig, waitForConfig } from '../config/index.js';
import logger from '../logger/index.js';
import { getPlugins } from '../plugins/index.js';
import docker from './docker.js';
import { initNetwork } from './network.js';
import { initTraefik } from './traefik.js';

// export default function
export async function initDocker() {
  await waitForConfig();

  logger.info('Initializing docker services...');
  // create exoframe network if needed
  const exoNet = await initNetwork();

  // get config
  const config = getConfig();

  // run init via plugins if available
  const plugins = getPlugins();
  logger.debug('Got plugins, running init:', plugins);
  for (const plugin of plugins) {
    // only run plugins that have init function
    if (!plugin.init) {
      continue;
    }

    const result = await plugin.init({ config, logger, docker });
    logger.debug('Initing traefik with plugin:', plugin.config.name, result);
    if (result && plugin.config.exclusive) {
      logger.info('Init finished via exclusive plugin:', plugin.config.name);
      return;
    }
  }

  // run traefik init
  await initTraefik(exoNet);
}
