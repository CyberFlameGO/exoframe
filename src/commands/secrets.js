// npm packages
const got = require('got');
const chalk = require('chalk');
const inquirer = require('inquirer');

// our packages
const {userConfig, isLoggedIn, logout} = require('../config');

exports.command = ['secret [cmd] [name] [value]'];
exports.describe = 'create, list or remove deployment secrets';
exports.builder = {
  cmd: {
    default: 'new',
    description: 'command to execute [new | ls | get | rm]',
  },
  name: {
    description: 'name of the secret',
  },
  value: {
    description: 'new value of the secret',
  },
};
exports.handler = async args => {
  if (!isLoggedIn()) {
    return;
  }

  // services request url
  const remoteUrl = `${userConfig.endpoint}/secrets`;
  // get command
  const {cmd} = args;
  // if remove or ls - fetch secrets from remote, then do work
  if (cmd === 'ls' || cmd === 'rm' || cmd === 'get') {
    const actions = {
      ls: 'Listing',
      rm: 'Removing',
      get: 'Getting',
    };
    console.log(chalk.bold(`${actions[cmd]} deployment secret${cmd === 'ls' ? 's' : ''} for:`), userConfig.endpoint);

    // get secrets from server
    // construct shared request params
    const options = {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${userConfig.token}`,
      },
      json: true,
    };
    // try sending request
    let secrets = [];
    try {
      const {body} = await got(remoteUrl, options);
      secrets = body.secrets;
    } catch (e) {
      // if authorization is expired/broken/etc
      if (e.statusCode === 401) {
        logout(userConfig);
        console.log(chalk.red('Error: authorization expired!'), 'Please, relogin and try again.');
        return;
      }

      console.log(chalk.red('Error getting deployment secrets:'), e.toString());
      return;
    }

    if (cmd === 'ls') {
      console.log(chalk.bold('Got saved secrets:'));
      console.log('');
      secrets.map(t =>
        console.log(`  > ${chalk.green(`@${t.name}`)} ${chalk.gray(`[${new Date(t.meta.created).toLocaleString()}]`)}`)
      );
      if (!secrets.length) {
        console.log('  > No deployment secrets available!');
      }
      return;
    }

    // get selected secret from args
    let selectedSecret = args.name;

    // if it's not provided - present user with selection
    if (!selectedSecret || !selectedSecret.length) {
      const prompts = [];
      prompts.push({
        type: 'list',
        name: 'selectedSecret',
        message: `Choose secret to ${cmd === 'get' ? 'get' : 'remove'}:`,
        choices: secrets.map(t => t.name),
      });
      ({selectedSecret} = await inquirer.prompt(prompts));
    }

    // if getting secret - ask user once more if he's sure
    if (cmd === 'get') {
      const {doGet} = await inquirer.prompt([
        {
          type: 'confirm',
          name: 'doGet',
          message: 'Get secret value? (will be shown in plain text)',
          default: false,
        },
      ]);

      if (!doGet) {
        console.log(chalk.red('Stopping!'), 'User decided not to read secret value..');
        return;
      }

      // get secrets from server
      // construct shared request params
      const options = {
        method: 'GET',
        headers: {
          Authorization: `Bearer ${userConfig.token}`,
        },
        json: true,
      };
      let secret;
      try {
        const {body} = await got(`${remoteUrl}/${selectedSecret}`, options);
        secret = body.secret;
      } catch (e) {
        // if authorization is expired/broken/etc
        if (e.statusCode === 401) {
          logout(userConfig);
          console.log(chalk.red('Error: authorization expired!'), 'Please, relogin and try again.');
          return;
        }

        console.log(chalk.red('Error getting deployment secret:'), e.toString());
        return;
      }

      console.log(chalk.bold('New secret generated:'));
      console.log('');
      console.log(`Name: ${secret.name}`);
      console.log(`Value: ${secret.value}`);
      console.log(`Date: ${new Date(secret.meta.created).toLocaleString()}`);

      return;
    }

    // construct shared request params
    const rmOptions = {
      method: 'DELETE',
      headers: {
        Authorization: `Bearer ${userConfig.token}`,
      },
      json: true,
      body: {
        secretName: selectedSecret,
      },
    };
    try {
      const {body, statusCode} = await got(remoteUrl, rmOptions);
      if (statusCode !== 204) {
        console.log(chalk.red('Error removing deployment secret!'), body.reason || 'Please try again!');
        return;
      }
      console.log(chalk.green('Deployment secret successfully removed!'));
    } catch (e) {
      // if authorization is expired/broken/etc
      if (e.statusCode === 401) {
        logout(userConfig);
        console.log(chalk.red('Error: authorization expired!'), 'Please, relogin and try again.');
        return;
      }

      console.log(chalk.red('Error removing secret:'), e.toString());
      return;
    }

    return;
  }

  console.log(chalk.bold('Generating new deployment secret for:'), userConfig.endpoint);

  let secretName = args.name;
  let secretValue = args.value;

  // if user haven't provided name and value - ask interactively
  if (!secretName || !secretValue) {
    // ask for secret name and value
    const prompts = [];
    prompts.push({
      type: 'input',
      name: 'secretName',
      message: 'Secret name:',
      validate: input => input && input.length > 0,
      filter: input => input.trim(),
    });
    prompts.push({
      type: 'input',
      name: 'secretValue',
      message: 'Secret value:',
      validate: input => input && input.length > 0,
      filter: input => input.trim(),
    });
    ({secretName, secretValue} = await inquirer.prompt(prompts));
  }

  // construct shared request params
  const options = {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${userConfig.token}`,
    },
    json: true,
    body: {
      secretName,
      secretValue,
    },
  };
  // try sending request
  try {
    const {body} = await got(remoteUrl, options);
    console.log(chalk.bold('New secret generated:'));
    console.log('');
    console.log(`Name: ${body.name}`);
    console.log(`Value: ${body.value}`);
    console.log('');
    console.log(chalk.green('DONE!'));
  } catch (e) {
    // if authorization is expired/broken/etc
    if (e.statusCode === 401) {
      logout(userConfig);
      console.log(chalk.red('Error: authorization expired!'), 'Please, relogin and try again.');
      return;
    }

    console.log(chalk.red('Error generating deployment secret:'), e.toString());
  }
};
