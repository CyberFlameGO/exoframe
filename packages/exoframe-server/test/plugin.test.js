/* eslint-env jest */
// npm packages
import getPort from 'get-port';
import { join } from 'path';
import { __load } from '../src/config/index.js';
import { secretsInited } from '../src/db/secrets.js';
import { initDocker } from '../src/docker/init.js';
import { start, startFromParams } from '../src/docker/start.js';
import { startServer } from '../src/index.js';
import { getPlugins, initPlugins } from '../src/plugins/index.js';
// our packages
import authToken from './fixtures/authToken.js';
// mock config for testing
jest.mock('../src/config', () => require('./__mocks__/config'));
jest.mock('../src/util', () => require('./__mocks__/util'));
jest.mock('./fixtures/plugins/node_modules/testplugin', () => require('./__mocks__/testplugin'), { virtual: true });
// switch config to plugins
__load('plugins');

// test project path
const testProjectPath = join(__dirname, 'fixtures', 'html-project');
// options base
const options = {
  method: 'GET',
  headers: {
    Authorization: `Bearer ${authToken}`,
  },
};

beforeAll(async (done) => {
  await secretsInited;
  await initPlugins();
  done();
});

test('Should init plugins', async (done) => {
  const plugins = getPlugins();
  expect(plugins.length).toEqual(1);

  const testPlugin = plugins[0];
  expect(testPlugin).toMatchSnapshot();

  done();
});

test('init', async (done) => {
  await initDocker();

  const plugins = getPlugins();
  const testPlugin = plugins[0];
  expect(testPlugin.init).toBeCalled();

  done();
});

test('start', async (done) => {
  await start({ image: 'test', username: 'admin', folder: testProjectPath, resultStream: {} });

  const plugins = getPlugins();
  const testPlugin = plugins[0];
  expect(testPlugin.start).toBeCalled();

  done();
});

test('startFromParams', async (done) => {
  await startFromParams({
    image: 'test',
    deploymentName: 'test-deploy',
    projectName: 'test-project',
    username: 'admin',
    backendName: 'back',
    frontend: 'front',
    hostname: 'host',
    restartPolicy: 'always',
  });

  const plugins = getPlugins();
  const testPlugin = plugins[0];
  expect(testPlugin.startFromParams).toBeCalled();

  done();
});

test('list', async (done) => {
  // start server
  const port = await getPort();
  const fastify = await startServer(port);

  const response = await fastify.inject({ ...options, url: '/list' });
  const result = JSON.parse(response.payload);

  // check response
  expect(result).toBeDefined();
  expect(response.statusCode).toEqual(200);

  const plugins = getPlugins();
  const testPlugin = plugins[0];
  expect(testPlugin.list).toBeCalled();

  await fastify.close();
  done();
});

test('logs', async (done) => {
  // start server
  const port = await getPort();
  const fastify = await startServer(port);

  const response = await fastify.inject({ ...options, url: '/logs/test' });
  const result = JSON.parse(response.payload);

  // check response
  expect(result).toBeDefined();
  expect(response.statusCode).toEqual(200);

  const plugins = getPlugins();
  const testPlugin = plugins[0];
  expect(testPlugin.logs).toBeCalled();

  await fastify.close();
  done();
});

test('remove', async (done) => {
  // start server
  const port = await getPort();
  const fastify = await startServer(port);

  const response = await fastify.inject({ ...options, method: 'POST', url: `/remove/test` });
  const result = JSON.parse(response.payload);

  // check response
  expect(result).toBeDefined();
  expect(response.statusCode).toEqual(200);

  const plugins = getPlugins();
  const testPlugin = plugins[0];
  expect(testPlugin.remove).toBeCalled();

  await fastify.close();
  done();
});
