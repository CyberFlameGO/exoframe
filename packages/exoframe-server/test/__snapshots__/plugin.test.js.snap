// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should init plugins 1`] = `
Object {
  "compose": [MockFunction],
  "config": Object {
    "exclusive": true,
    "name": "test",
  },
  "init": [MockFunction],
  "list": [MockFunction],
  "logs": [MockFunction],
  "remove": [MockFunction],
  "start": [MockFunction],
  "startFromParams": [MockFunction],
}
`;
